### ISSUE TITLE
C++: bug when using function as parameter

### ISSUE THREAD

**Issue by sehraf** [doxygen/doxygen#7236]
Hi,

i've the following declaration:
```C++
	/**
	 * @brief Request remote files search
	 * @jsonapi{development}
	 * @param[in] matchString string to look for in the search
	 * @param multiCallback function that will be called each time a search
	 * result is received
	 * @param[in] maxWait maximum wait time in seconds for search results
	 * @return false on error, true otherwise
	 */
	virtual bool turtleSearchRequest(
	        const std::string& matchString,
	        const std::function<void (const std::list<TurtleFileInfo>& results)>& multiCallback,
	        rstime_t maxWait = 300 ) = 0;
```
Notice the function parameter `multiCallback`.

Previously (1.8.15) this resulted in this (correct) XML output:
```XML
   <param>
      <type>const std::function&lt; void(const std::list&lt; <ref refid="structTurtleFileInfo" kindref="compound">TurtleFileInfo</ref> &gt; &amp;results)&gt; &amp;</type>
      <declname>multiCallback</declname>
   </param>
```

Now (1.8.16) this results in this (broken) output:
```XML
   <param>
      <type>const std::function&lt; void(const std::list&lt; <ref refid="structTurtleFileInfo" kindref="compound">TurtleFileInfo</ref> &gt; &amp;</type>
      <declname>results</declname>
   </param>
```

As you can see, the `declname` entry now has the wrong name and `type` is  cut of.
Doxygen is used with this [config](https://github.com/RetroShare/RetroShare/blob/master/jsonapi-generator/src/jsonapi-generator-doxygen.conf) (with additional OUTPUT_DIRECTORY and INPUT set)

**Comment by albert-github** [comment]
Please add the configuration file to the issue as external links might not be persistent.
Best is to:
- attach a, small, self contained example (source+configuration file in a tar or zip) that allows us to reproduce the problem.

Worthwhile to mention is also that with the 1.8.16 version there are a number of warnings:
```
../aa.h:3: warning: argument 'multiCallback' of command @param is not found in the argument list of turtleSearchRequest(const std::string &matchString, const std::function< void(c
onst std::list< TurtleFileInfo > &results) &=0
../aa.h:3: warning: argument 'maxWait' of command @param is not found in the argument list of turtleSearchRequest(const std::string &matchString, const std::function< void(const s
td::list< TurtleFileInfo > &results) &=0
../aa.h:12: warning: The following parameters of turtleSearchRequest(const std::string &matchString, const std::function< void(const std::list< TurtleFileInfo > &results) &=0 are
not documented:
  parameter 'results'
```

**Comment by sehraf** [comment]
Thanks for the quick response!

I've attached a reproducer.
The errors you mention fit the output. The parsing process seems to stall after `&results`, hence neither `multiCallback` nor `maxWait` can be found! The errors don't exist with 1.8.15.

I didn't even notice previously that `maxWait` is missing, too. When looking at the diff of the resulting XML output, you can see that the function is kind of cut of:
```diff
         <param>
-          <type>const std::function&lt; void(const std::list&lt; TurtleFileInfo &gt; &amp;results)&gt; &amp;</type>
-          <declname>multiCallback</declname>
-        </param>
-        <param>
-          <type>rstime_t</type>
-          <declname>maxWait</declname>
-          <defval>300</defval>
+          <type>const std::function&lt; void(const std::list&lt; TurtleFileInfo &gt; &amp;</type>
+          <declname>results</declname>
         </param>
```

[doxygen-reproducer.zip](https://github.com/doxygen/doxygen/files/3555623/doxygen-reproducer.zip)

**Comment by albert-github** [comment]
Problem (at least the appearance of the warnings) looks like to be caused by the fix: 
Improve handling < for expressions inside template argument defaults(44c7ab97f03ef94d5965d0419def4931d0aa61eb )

**Comment by doxygen** [comment]
Look like a logic bug. Please check if the referenced commit fixes the problem.

**Comment by sehraf** [comment]
Hey, i can confirm that 6a04051970cf66330a0a3bce49abf9f4f60b6d17 fixes the problem! (Tested with 6fe6275cd66c7842d6817fd4b8ea0a47b2b5df20, one commit later, current master)

Thanks for the quick fix!
Is there any (rough) release eta?

**Comment by albert-github** [comment]
Please don't close the issue as this will be done at the moment of a release

**Comment by doxygen** [comment]
@sehraf Next release is probably around the end of this year.

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.8.17.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
